### PR TITLE
Implement XEP-0198: Stream Management (client only)

### PR DESCRIPTION
**Pull request by olesalscheider**
This PR implements the client side of XEP-0198: Stream Management.

It supports resumption but does not yet indicate if the stream was resumed. This should be changed at some point so that the client does not again send presence information or request the roster.

### CODE REVIEW

**Comment by olesalscheider** [comment]
Are there any news on this?
This PR fixes https://github.com/qxmpp-project/qxmpp/issues/66

**Comment by Kaffeine** [comment]
@olesalscheider  Would you mind to rebase the PR on master?
Looks like it would solve the problem with travis-ci and may be then we would have this change intergrated or, at least reviewed.

**Comment by olesalscheider** [comment]
@Kaffeine Sure, done.

**Comment by renini** [comment]
Is there any news if this will be implemented?

File: src/base/QXmppStream.cpp [hunk 1]
```diff
@@ -262,4 +287,95 @@ void QXmppStream::_q_socketReadyRead()
...
+        if (!d->unackedStanzas.empty()) {
+            for (QMap<unsigned, QByteArray>::iterator it = d->unackedStanzas.begin(); it != d->unackedStanzas.end(); ++it)
+                sendData(it.value());
+            sendAckRequest();
+        }
+    }
+}
 
+/// Returns the sequence number of the last incoming stanza (XEP-0198).
+unsigned QXmppStream::lastIncomingSeqNo() const
```
Comment (jlaine) [inline src/base/QXmppStream.cpp]: I am not fond of abreviations, I would prefer "lastIncomingSequenceNumber()"

File: src/base/QXmppStream.cpp [hunk 2]
```diff
@@ -262,4 +287,95 @@ void QXmppStream::_q_socketReadyRead()
...
+{
+    return d->lastIncomingSeqno;
+}
+
+/// Sets the last acknowledged sequence number for outgoing stanzas (XEP-0198).
+void QXmppStream::setAckedSeqNo(unsigned seqno)
```
Comment (jlaine) [inline src/base/QXmppStream.cpp]: ditto, no abreviations

File: src/base/QXmppStreamFeatures.cpp [hunk 3]
```diff
@@ -94,6 +95,16 @@ void QXmppStreamFeatures::setTlsMode(QXmppStreamFeatures::Mode mode)
     m_tlsMode = mode;
 }
 
+QXmppStreamFeatures::Mode QXmppStreamFeatures::streamManagementMode() const
```
Comment (jlaine) [inline src/base/QXmppStreamFeatures.cpp]: Please document this API

File: src/base/QXmppStreamFeatures.cpp [hunk 4]
```diff
@@ -94,6 +95,16 @@ void QXmppStreamFeatures::setTlsMode(QXmppStreamFeatures::Mode mode)
...
+{
+    return m_streamManagementMode;
+}
+
+void QXmppStreamFeatures::setStreamManagementMode(QXmppStreamFeatures::Mode mode)
```
Comment (jlaine) [inline src/base/QXmppStreamFeatures.cpp]: Ditto, please document this API

File: src/client/QXmppOutgoingClient.h [hunk 5]
```diff
@@ -102,6 +105,9 @@ private slots:
 private:
     void sendNonSASLAuth(bool plaintext);
     void sendNonSASLAuthQuery();
+    void sendBind();
```
Comment (jlaine) [inline src/client/QXmppOutgoingClient.h]: Move these to QXmppOutgoingClientPrivate

**Comment by olesalscheider** [comment]
Is there still something that I should change?
We would like to see this in QXmpp for telepathy-nonsense.

**Comment by jlaine** [comment]
Thanks for your contribution ! By the way as you have noticed, we are a bit short staffed, any interest in helping maintain QXmpp on a regular basis?

**Comment by olesalscheider** [comment]
@jlaine Yes, I would be willing to help maintaining QXmpp. It is quite important for telepathy-nonsense.

File: src/base/QXmppStreamManagement.cpp [hunk 6]
```diff
@@ -0,0 +1,325 @@
...
+    m_previd = element.attribute("previd");
+}
+
+void QXmppStreamManagementResume::toXml(QXmlStreamWriter *writer) const
+{
+    writer->writeStartElement("resume");
+    writer->writeAttribute("h", QString::number(m_h));
+    writer->writeAttribute("previd", m_previd);
+    writer->writeEndElement();
+}
```
Comment (lnjX) [inline src/base/QXmppStreamManagement.cpp:193]: missing stream management `xmlns`

**Pull request merged**